### ISSUE TITLE
Cpputest 3.7 fails to build in IAR

### ISSUE THREAD

**Issue by imr** [cpputest/cpputest#664]
I have built cpputest 3.6 in IAR, but 3.7 and following versions do not. I have a project directory "Eval" and cpputest in a subdirectory. All other settings are the same as found at https://cpputest.github.io/stories.html#iar  Compiler output below:

Updating build tree... 
CommandLineTestRunner.cpp  
CommandLineArguments.cpp  
JUnitTestOutput.cpp  
MemoryLeakDetector.cpp  
MemoryLeakWarningPlugin.cpp  
SimpleString.cpp  
TestFilter.cpp  
TestFailure.cpp  
TestHarness_c.cpp  
TestMemoryAllocator.cpp  
TestOutput.cpp  
TestPlugin.cpp  
TestResult.cpp  
TestRegistry.cpp  
UtestPlatform.cpp  
Error[Pe147]: declaration is incompatible with "extern "C" int (*PlatformSpecificSetJmp)(void (*)(void *), void *)" (declared at line 45 of "C:\<project location>\Eval\EWARM\..\cpputest\include\CppUTest/PlatformSpecificFunctions_c.h") C:\<project location>\Eval\cpputest\src\Platforms\Iar\UtestPlatform.cpp 47 
Error[Pe147]: declaration is incompatible with "extern "C" void (*PlatformSpecificLongJmp)()" (declared at line 46 of "C:\<project location>\Eval\EWARM\..\cpputest\include\CppUTest/PlatformSpecificFunctions_c.h") C:\<project location>\Eval\cpputest\src\Platforms\Iar\UtestPlatform.cpp 58 
Error[Pe147]: declaration is incompatible with "extern "C" void (*PlatformSpecificRestoreJumpBuffer)()" (declared at line 47 of "C:\<project location>\Eval\EWARM\..\cpputest\include\CppUTest/PlatformSpecificFunctions_c.h") C:\<project location>\Eval\cpputest\src\Platforms\Iar\UtestPlatform.cpp 64 
Error[Pe147]: declaration is incompatible with "void (*PlatformSpecificRunTestInASeperateProcess)(UtestShell *, TestPlugin *, TestResult *)" (declared at line 35 of "C:\<project location>\Eval\EWARM\..\cpputest\include\ C:\<project location>\Eval\cpputest\src\Platforms\Iar\UtestPlatform.cpp 69 
CppUTest/PlatformSpecificFunctions.h") 
Error[Pe147]: declaration is incompatible with "extern "C" int (*PlatformSpecificVSNprintf)(char *, std::size_t, char const *, __va_list)" (declared at line 54 of "C:\<project location>\Eval\EWARM\..\cpputest\include\ C:\<project location>\Eval\cpputest\src\Platforms\Iar\UtestPlatform.cpp 103 
CppUTest/PlatformSpecificFunctions_c.h") 
Error[Pe147]: declaration is incompatible with "extern "C" void * (*PlatformSpecificFOpen)(char const *, char const *)" (declared at line 64 of "C:\<project location>\Eval\EWARM\..\cpputest\include\ C:\<project location>\Eval\cpputest\src\Platforms\Iar\UtestPlatform.cpp 108 
CppUTest/PlatformSpecificFunctions_c.h") 
Error[Pe147]: declaration is incompatible with "extern "C" void (*PlatformSpecificFPuts)(char const *, void *)" (declared at line 65 of "C:\<project location>\Eval\EWARM\..\cpputest\include\CppUTest/PlatformSpecificFunctions_c.h") C:\<project location>\Eval\cpputest\src\Platforms\Iar\UtestPlatform.cpp 115 
Error[Pe147]: declaration is incompatible with "extern "C" void (*PlatformSpecificFClose)(void *)" (declared at line 66 of "C:\<project location>\Eval\EWARM\..\cpputest\include\CppUTest/PlatformSpecificFunctions_c.h") C:\<project location>\Eval\cpputest\src\Platforms\Iar\UtestPlatform.cpp 121 
Error[Pe147]: declaration is incompatible with "extern "C" void (*PlatformSpecificFlush)()" (declared at line 69 of "C:\<project location>\Eval\EWARM\..\cpputest\include\CppUTest/PlatformSpecificFunctions_c.h") C:\<project location>\Eval\cpputest\src\Platforms\Iar\UtestPlatform.cpp 126 
Error[Pe147]: declaration is incompatible with "extern "C" int (*PlatformSpecificPutchar)(int)" (declared at line 68 of "C:\<project location>\Eval\EWARM\..\cpputest\include\CppUTest/PlatformSpecificFunctions_c.h") C:\<project location>\Eval\cpputest\src\Platforms\Iar\UtestPlatform.cpp 130 
Error[Pe147]: declaration is incompatible with "extern "C" void *(*PlatformSpecificMalloc)(std::size_t)" (declared at line 72 of "C:\<project location>\Eval\EWARM\..\cpputest\include\CppUTest/PlatformSpecificFunctions_c.h") C:\<project location>\Eval\cpputest\src\Platforms\Iar\UtestPlatform.cpp 135 
Error[Pe147]: declaration is incompatible with "extern "C" void *(*PlatformSpecificRealloc)(void *, std::size_t)" (declared at line 73 of "C:\<project location>\Eval\EWARM\..\cpputest\include\CppUTest/PlatformSpecificFunctions_c.h") C:\<project location>\Eval\cpputest\src\Platforms\Iar\UtestPlatform.cpp 140 
Error[Pe147]: declaration is incompatible with "extern "C" void (*PlatformSpecificFree)(void *)" (declared at line 74 of "C:\<project location>\Eval\EWARM\..\cpputest\include\CppUTest/PlatformSpecificFunctions_c.h") C:\<project location>\Eval\cpputest\src\Platforms\Iar\UtestPlatform.cpp 145 
Error[Pe147]: declaration is incompatible with "extern "C" void *(*PlatformSpecificMemCpy)(void *, void const *, std::size_t)" (declared at line 75 of "C:\<project location>\Eval\EWARM\..\cpputest\include\ C:\<project location>\Eval\cpputest\src\Platforms\Iar\UtestPlatform.cpp 150 
CppUTest/PlatformSpecificFunctions_c.h") 
Error[Pe147]: declaration is incompatible with "extern "C" void *(*PlatformSpecificMemset)(void *, int, std::size_t)" (declared at line 76 of "C:\<project location>\Eval\EWARM\..\cpputest\include\CppUTest/PlatformSpecificFunctions_c.h") C:\<project location>\Eval\cpputest\src\Platforms\Iar\UtestPlatform.cpp 155 
Error[Pe147]: declaration is incompatible with "extern "C" double (*PlatformSpecificFabs)(double)" (declared at line 57 of "C:\<project location>\Eval\EWARM\..\cpputest\include\CppUTest/PlatformSpecificFunctions_c.h") C:\<project location>\Eval\cpputest\src\Platforms\Iar\UtestPlatform.cpp 160 
Error while running C/C++ Compiler 
Utest.cpp  
 
Total number of errors: 16 
Total number of warnings: 0

**Comment by imr** [comment]
I looked into this more and the commit that breaks it is 59ca1fa3d74ca6af562bf3cbbe3f0faab39e71a6

**Comment by arstrube** [comment]
That makes sense, because no-one updated Iar/UtestPlatform.cpp to match the changes in that commit. Would you be so kind to add those, taking Gcc/UtestPlatform.cpp as a model? I could do it, but am hesitant to do so as I don't have the setup to test-compile with IAR.

(Basically, several functions were changed to a function-pointer based interface, so they can be easily swapped out for testing).

**Comment by imr** [comment]
How does the compiler get the platform? Gcc, GccNoStdC, etc. I can append IAR to the front of these functions, but the compiler complains that they cannot be found when compiling the tests.

**Comment by arstrube** [comment]
It finds them by *magic* ;-)

It's not quite as simple as that, although almost : -). I'll put it together for you if I can spare the time.
You need to edit Platforms/IAR/UtestPlatform.cpp. It needs to *formally* look like /Gcc/UtestPlatform.cpp, but of course keep its own contents of any function definitions.

**Comment by arstrube** [comment]
Here it is: [UtestPlatform.cpp](https://github.com/arstrube/cpputest/blob/c38cf192d0b096c8614742d2920b4a02420cb1e0/src/Platforms/Iar/UtestPlatform.cpp)
I did not have a chance to run it past IAR, but at least it does compile with Gcc.
If anything doesn't work with IAR, let me kow.

**Comment by arstrube** [comment]
Come to think of it, it might be more appropriate to call the SeparateProcess and fork stuff `Dummy....` rather than `Iar...` or `...Implementation`.

**Comment by imr** [comment]
This is close to working. Using the master branch at commit 7d3451f1 and your updated file, cpputest compiles as well as tests. When I run the compiled tests I get an error on line 170 of UtestPlatform.cpp: PlatformSpecificMutexCreate is not implemented.

**Comment by arstrube** [comment]
Would you be so kind as to post the entire error? I would expect several tests to fail, because some functions are not implemented for Iar, and there probably needs to be a matching `#ifdef`in UtestPlatformTest.cpp. Is there anything your compiler defines (such as __IAR__ or whatever) for your platform, that we can use to determine which test should be run?

**Comment by imr** [comment]
Stepping through with the debugger I see that it does indeed fail in the DummyMutexCreate function. If the failure on line 170 is commented out, the tests are run as expected. I am not using the tests on the cpputest codebase, just simple ones like the ones described at https://cpputest.github.io/stories.html#iar The entire error message is
Unknown file:0: Failure in TEST(
    NOTE: Assertion happened without being in a test run (perhaps in main?), Something is very wrong. Check this assertion and fix)
C:\<project location>\cpputest\src\Platforms\Iar\UtestPlatform.cpp:170: error: PlatformSpecificMutexCreate is not implemented

**Comment by arstrube** [comment]
Can you perhaps check my latest pull request, whether it fixes this? The failure is unusual in that it does not occur inside a test run (which it should).

**Comment by arstrube** [comment]
> I am not using the tests on the cpputest codebase, just simple ones like the ones described at https://cpputest.github.io/stories.html#iar
Now, here is a clue. The big question is, how come this function is called at all, outside of a test of all places? 
The change here is that the attempt now causes a failure (which it should), whereas previously, it would just have produce a printf() message and run the test normally. Still, very strange. Any "evidence" would be helpful.

**Comment by arstrube** [comment]
> C:<project location>\cpputest\src\Platforms\Iar\UtestPlatform.cpp:170: error: PlatformSpecificMutexCreate is not implemented

Oh, now I remember. The `SimpleMutex` was added by another contributor. It uses `PlatformSpecificMutexCreate` indirectly, I believe, and it is used by CppUTest itself....

**Comment by imr** [comment]
I tried the latest and it does not quite work. The call stack when line 170 of UtestPlatform.CPP is encountered is as follows:
DummyMutexCreate()
SimpleMutex::SimpleMutex()
MemoryLeakDetector::MemoryLeakDetector(class MemoryLeakFailure *)
MemoryLeakWarningPlugin::getGlobalDetector()
MemoryLeakWarningPlugin::MemoryLeakWarningPlugin(class SimpleString const &, class MemoryLeakDetector *)
CommandLineTestRunner::RunAllTests(int, char const * *)
CommandLineTestRunner::RunAllTests(int, char * *)
main(int, char * *)

I didn't add anything that would start the memory leak detector. Is it automatic? Can it run without mutex?

**Comment by arstrube** [comment]
Yeah, I found that too. The thing is, it needs to be implemented. It is used by the memory leak warning feature, as you discovered. Shouldn't be too hard. 

You could use this: https://www.iar.com/Support/resources/articles/beyond-volatile/ as a starting point.

**Comment by arstrube** [comment]
I don't know whether you really need that. An empty implementation might be in order (just remove the FAIL()'s

**Comment by arstrube** [comment]
My guess would be the MemoryLeakDetection can run with a Mutex that does nothing, because there aren't any threads executing in parallel, right @basvodde ?

**Comment by arstrube** [comment]
BTW you can disable the memory leak detector, but I would do that. Rather throw out the FAIL lines.

**Comment by arstrube** [comment]
Sorry for the monologue :)

You could take the information from above mentioned article and do something this:
```c++
static unsigned char mutex_;

static PlatformSpecificMutex IARMutexCreate(void)
{
    return &mutex_;
} 

static void DummyMutexLock(PlatformSpecificMutex mutex)
{
    *mutex = __save_interrupt();
    __disable_interrupt(); 
}

static void DummyMutexUnlock(PlatformSpecificMutex mutex)
{
     __restore_interrupt(*mutex);
}

static void DummyMutexDestroy(PlatformSpecificMutex)
{
}
```
Let us know whether you find this more useful (does it work?) or rather the empty implementation. I can then change my pull request to match.

**Comment by arstrube** [comment]
errr... forgot the type casts:
```c++
static unsigned char mutex_;

static PlatformSpecificMutex IarMutexCreate(void)
{
    return (PlatformSpecificMutex)&mutex_;
}

static void IarMutexLock(PlatformSpecificMutex mutex)
{
    *(unsigned char*) mutex = __save_interrupt();
    __disable_interrupt();
}

static void IarMutexUnlock(PlatformSpecificMutex mutex)
{
     __restore_interrupt(*(unsigned char*) mutex);
}

static void IarMutexDestroy(PlatformSpecificMutex)
{
}

PlatformSpecificMutex (*PlatformSpecificMutexCreate)(void) = IarMutexCreate;
void (*PlatformSpecificMutexLock)(PlatformSpecificMutex) = IarMutexLock;
void (*PlatformSpecificMutexUnlock)(PlatformSpecificMutex) = IarMutexUnlock;
void (*PlatformSpecificMutexDestroy)(PlatformSpecificMutex) = IarMutexDestroy;
```

**Comment by imr** [comment]
Removing the code in the mutex functions allows the simple tests to run.

The interrupt functions don't compile, ie. __save_interrupt. I believe that part of the article is in reference to the AVR architecture. I compiled the code that is to be under test, cpputest, and the simple tests to target a Cortex M4 device. Maybe I need a more general ARM architecture for compiling the tests.

**Comment by arstrube** [comment]
That's a pity. The article mentiones __get_CPSR(), but is rather incomplete with regards to ARM. It also doesn't mention any headers one might need to include

**Comment by arstrube** [comment]
I think probably using a dummy implementation makes more sense anyway. And yes, I would think the memory leak detector can run with the dummy mutex.

Can we close the issue? Or is anything still amiss?

**Comment by imr** [comment]
Yes, this issue can be closed. If I can get a replacement for the dummy code together, I will submit a separate pull request.
